core/src/com/unciv/logic/civilization/managers/ThreatManager.kt

```diff
@@ -34,69 +34,97 @@ class ThreatManager(val civInfo: Civilization) {
+                return if (takeLargerValues) notFoundDistance else maxDist
             }
+
             // Only search the tiles that we haven't searched yet
             minDistanceToSearch = (tileData.distanceSearched + 1).coerceAtLeast(1)
         }
 
+        if (tileData != null && tileData.tilesWithEnemies.isNotEmpty()) throw IllegalStateException("There must be no elements in tile.data.tilesWithEnemies at this point")
+        val tilesWithEnemyAtDistance: MutableList<Pair<Tile,Int>> = mutableListOf()
         // Search for nearby enemies and store the results
         for (i in minDistanceToSearch..maxDist) {
             for (searchTile in tile.getTilesAtDistance(i)) {
                 if (doesTileHaveMilitaryEnemy(searchTile)) {
-                    // We have only completely searched a radius of  i - 1 
-                    distanceToClosestEnemyTiles[tile] = ClosestEnemyTileData(i - 1, i, searchTile)
-                    return i
+                    tilesWithEnemyAtDistance.add(Pair(searchTile, i))
                 }
             }
+            if (tilesWithEnemyAtDistance.isNotEmpty()) {
+                distanceToClosestEnemyTiles[tile] = ClosestEnemyTileData(i, tilesWithEnemyAtDistance)
+                return i
+            }
         }
-        distanceToClosestEnemyTiles[tile] = ClosestEnemyTileData(maxDist, null, null)
+        distanceToClosestEnemyTiles[tile] = ClosestEnemyTileData(maxDist, mutableListOf())
         return notFoundDistance
     }
 
     /**
      * Returns all tiles with enemy units on them in distance.
+     * Every tile is guaranteed to have an enemy.
      * May be quicker than a manual search because of caching.
      * Also ends up calculating and caching [getDistanceToClosestEnemyUnit].
      */
     fun getTilesWithEnemyUnitsInDistance(tile: Tile, maxDist: Int): MutableList<Tile> {
         val tileData = distanceToClosestEnemyTiles[tile]
-        
-        // Shortcut, we don't need to search for anything
+
+        val tilesWithEnemies: MutableList<Tile> = mutableListOf()
+        val tileDataTilesWithEnemies: MutableList<Pair<Tile,Int>> = if (tileData?.tilesWithEnemies != null) tileData.tilesWithEnemies else mutableListOf()
+
+
+        if (tileData != null && tileData.distanceSearched >= maxDist) {
+            // Add all tiles that we have previously found
+            var index = 0
+            while (tileDataTilesWithEnemies.size > index) {
+                val tileWithDistance = tileDataTilesWithEnemies[index]
+                if (tileWithDistance.second > maxDist)
+                    return tilesWithEnemies
+                if (doesTileHaveMilitaryEnemy(tileWithDistance.first)) {
+                    tilesWithEnemies.add(tileWithDistance.first)
+                } else {
+                    tileDataTilesWithEnemies.removeAt(index)
+                    continue
+                }
+                index++
+            }
+        }
```

Comment:
This feels needlessly confusing and I'm having a hard time trying to formulate what is wrong.

This part should be *converting the tileData to a list of tiles within range that have enemies*, right? 
So
A. In what case would the cached tileDataTilesWithEnemies *not* have an enemy? If we got rid of it with a previous unit? That should be explicit, this is very roundabout!
B. What we actually care about *here* is the tilesWithEnemies!

So I would formulate this as:

```kotlin
if (...)
 return tileDataTilesWithEnemies.filter {it.second <= maxDist}
.filter { doesTileHaveMilitaryEnemy(tileWithDistance.first }
```

If you want to update the tileDataTilesWithEnemies to *remove* those tiles that we determined no longer have units, that's fine, you can do that as a separate step, but not at expense of readability.

Reply:
Do you find the version with the mutableListIterator to be more readable? Functional programming doesn't seem to be efficient with having to deal with two things at once (which goes against the whole point of functional programming). We could loop through it twice but then we also won't be using the sorted property of the list.

Reply:
For context, I plan on running this section of code indirectly a lot more in the future. There are also some other performance improvements that I am investigating will make the code bigger and less readable. If someone wants to make a change in this area of the code, they should probably have a much better understanding of how it works compared to code in other parts of Unciv. If we need to quickly change a feature we probably shouldn't be looking to modify this code, especially if it will be used in many places.

Ideally, each military unit would call a variant of `getTilesWithEnemiesInDistance(4)` or `gTWED(4)` (60 tiles to search) to look around the surrounding area to see if they should continue attacking, hold or retreat. Preferably also on the tiles melee units want to attack move to as well to make sure they are safe enough.
If we want to check if a city is being surrounded by neutral units, we might also want to call `gTWED(6)` (126 tiles to search) on cities to find those units.
Air units are calling `gTWED(5, 8, 20)` (90, 216, 1260 tiles to search) every turn to attack.
Air units attack before all ground units, which is great because it would overshadow the city check that they are in and we could use the information gathered from the larger checks to speed up the smaller checks around it.

An example of where ending the search early is very helpful: a stealth bomber calls `gTWED(20)` and finds some enemies, then we call `gTWED(6)` of smaller for any reason. The original range of 20 call has 1134 tiles that the smaller call does not overlap with. If we don't end the search early, each smaller call would have to search through all previously found enemies on the additional 1134 tiles (note that each tile must either have a city or be visible to count).